app/controllers/voucher_adjustments_controller.rb

```diff
@@ -74,4 +79,10 @@ def render_error
   def voucher_params
     params.require(:order).permit(:voucher_code)
   end
+
+  # Clear payments and payment fees, to not affect voucher adjustment calculation
+  def clear_payments
+    @order.all_adjustments.payment_fee.destroy_all
+    @order.payments.clear
```

Comment:
My understanding is that payment are processed once we confirm the order, see below: 
https://github.com/openfoodfoundation/openfoodnetwork/blob/ad37c401e5866f0ae2e432c37076885a004f2a1f/app/controllers/split_checkout_controller.rb#L115-L126

and once the order is confirmed then the state machine will finalise the order. So If I understand correctly we shouldn't have any completed payments when navigating between checkout steps. 
Also if the checkout is restarted, the the user will be going through the payment step again and will be able to (re) choose the payment method. 

I added this to fix a specific scenario where the user goes through the checkout till the confirmation step, then goes back to the payment step and add a voucher covering the order amount. In this case we need to remove any existing payment/payment fee as now there is no payment required. 

Would we have an already completed payment if we face a stock conflict and have to restart the checkout ? You would think that we would check the stock situation before taking a payment. 
From what I can see, we would be redirected to cart if there is a stock issue when navigation checkout steps:
https://github.com/openfoodfoundation/openfoodnetwork/blob/ad37c401e5866f0ae2e432c37076885a004f2a1f/app/controllers/concerns/checkout_callbacks.rb#L34
https://github.com/openfoodfoundation/openfoodnetwork/blob/ad37c401e5866f0ae2e432c37076885a004f2a1f/app/controllers/concerns/checkout_callbacks.rb#L66-L70
I couldn't see any other checks so maybe I am missing something here.

Anyway, I'll apply @dacook suggestion to be on the safer side.

Reply:
> Would we have an already completed payment if we face a stock conflict and have to restart the checkout ? You would think that we would check the stock situation before taking a payment.
> From what I can see, we would be redirected to cart if there is a stock issue when navigation checkout steps

Of course we check the stock before we take a payment but there's always room for race conditions. We often process several checkouts in the same second and it can happen that two people try to checkout the same last item at the same time. It's really hard to predict what exactly happens because it depends on the timing and where the error is raised.

Actually, the CurrentOrderLocker should prevent this. So maybe it's not a problem any more. I'm just scared that this race condition will come back to haunt us... :fearful: